### PR TITLE
Blog post for how Pulumi can be used to automate member account creation in AWS Organizations

### PR DESCRIPTION
**Pull request by praneetloke**
🛑 This blog post requires the sources from the yet-to-be merged https://github.com/pulumi/examples/pull/1330.

I am not a huge fan of the meta image for this post but it's the best I could come up with right now. So please let me know if you have any other ideas for it. I'd be happy to update the file as long as it's within my drawio skills 😄

### CODE REVIEW

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (praneetloke) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:16]: I have to admit I am not very fond of this opener a whole lot.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (praneetloke) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:3]: I picked next Monday as the date but based on when this can be reviewed/merged, I'll update this.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (cnunciato) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:88]: Also here -- might be good to just mention briefly that these (this one and `AutomationRole`) are AWS built-ins.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (cnunciato) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:150]: This is cool, but feels somewhat orthogonal to the content of the post, which is about account creation. Totally fine to leave it in -- I just figured I'd mention it as something to consider if you were thinking about making any revisions.

(Out of curiosity, though, if you do leave it in, it might be good to mention how it works, and sprinkle in a link or two to the AWS docs. For example, I sort of assumed I'd see something about what kinds of "things" would be backed up (DB, etc., as you mention), but that doesn't seem to be called out in the code.)

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (cnunciato) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:259]: Might be good to explain what `costCenter` refers to.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (cnunciato) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:281]:
```suggestion
This post demonstrates how you could use the Pulumi programming model to encapsulate and account for the hierarchical organization structure that your team desires or is required to have for any number of reasons. While an enterprise organization might be interested in this sort of a setup to ensure multiple teams in the organization operate in a similar way, smaller teams should also strongly consider organizing their AWS accounts in a way that enforces the principle of least privileged access. Especially, if you are looking to get SOC2 certified and need to ensure strict isolation of resources that hold any kind of real-world customer data.
```

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (cnunciato) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:16]: I think this is fine!

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (cnunciato) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:283]: I'd really like to see what this looks like when it runs! Partly because as you suggest, many users won't have the appropriate permissions to actually deploy it. If there were some way to show what happens after a successful run, it'd really bring it all home I think. Up to you though.

**Review by cnunciato — APPROVED** [review]
This is a great post -- well done! Nothing need block. cc @scottslowe if you'd like another set of eyes, but to me this is great. My only request (aside from a few suggestions) would be to show a bit about what it looks like when you deploy and actually use this.

**Comment by scottslowe** [comment]
@cnunciato @aaronkao We should slot this into our content cadence schedule.

@praneetloke Not that you need it (if Christian says it's good then it's good), but I'm getting ready to review for another set of eyes.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (scottslowe) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:12]:
```suggestion
In an enterprise organization, an IT self-service "vending machine" allows employees to quickly and easily request and receive access to pre-approved cloud resources. Behind the scenes, Pulumi programs may orchestrate any of the requisite resources. We will look at an example of using Pulumi to create an AWS child account, within an AWS Organization.
```

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (scottslowe) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:96]:
```suggestion
Next, we'll create the IAM user that we’ll use for automation purposes. Again, you don’t need to do this if you plan to use OIDC federation for your CI/CD service.
```

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (scottslowe) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:150]: If I'm understanding correctly, I _think_ the idea here is to also show the "other things" that need to be done when creating accounts within an AWS Organization. If that's the case, I'd suggest some sort of transition here ("In addition to creating and deleting accounts, most teams will also want to make sure that resources are backed up, tagging is enforced...").

**Review by scottslowe — COMMENTED** [review]
Overall, great post! In addition to a +1 on all of @cnunciato's comments, I have a few suggested changes as well that I think would improve the post.

**Comment by praneetloke** [comment]
@cnunciato @scottslowe thank you for spending the time to review this. I am looking at the comments right now. I'll follow-up when it's ready for another review.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:24]: I'll reword "developer accounts" here. What I meant was the actual users of the AWS account who are developers. Not the development AWS accounts. Haha! I see how confusing this sentence is. Even I did a double-take reading this. 😄

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:283]: Yes definitely agree. Do you think a GIF or a screenshot of the Pulumi plan would be good here?

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:46]: Yes definitely. The `organization` variable's value depends on whether or not the user has already created the organization in their root account. But I'll include that snippet.

Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:46]: Ah yes. Looks like all code snippets in this post are somehow using three-space indentation. I must have miscounted them 😄 as I had to adjust these manually after a copy/paste from a Google Doc. I'll fix them.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:88]: This role and the `AutomationRole` are roles that the app creates. In the actual example I have these in two variables. I replaced the variable references with hard-coded strings for the sake of keeping things simple in the code snippet in this post. The relevant lines in the actual example: https://github.com/pulumi/examples/pull/1330/files#diff-b8897f56ce67084e1d7fd95fb1af74032ef404fee05abbb4ab013340f384b6a6R18-R19

Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:88]: But I'll add a comment that these are roles created by the component resource for account permissions.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:20]: Done. I've updated the paragraph before this slightly to do just that. Let me know what you think.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:176]: @cnunciato

> This is cool, but feels somewhat orthogonal to the content of the post, which is about account creation. Totally fine to leave it in -- I just figured I'd mention it as something to consider if you were thinking about making any revisions.

It's basically what @scottslowe mentions in his comment above. I think the suggestion to add a transition is a good one. I'll update this.

Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:176]: I've updated this section and the next. Let me know what you think.

**Comment by praneetloke** [comment]
@cnunciato @scottslowe I've resolved all of the suggestions. Can you please have another look? There are a couple of conversation threads that I didn't resolve just so it's easier to see the areas where a follow-up is required.

I've also got updates to the example [PR](https://github.com/pulumi/examples/pull/1330) as well and getting ready to push those up too.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (scottslowe) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:315]: An animated GIF that shows part of the `pulumi up` output would be cool, I think.

**Review by scottslowe — APPROVED** [review]
Just the one comment regarding adding an animated GIF (or screenshot) to help illustrate what it looks like when you run this, but we won't need to block on this.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (praneetloke) [reply themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:363]: I've expanded the outro a bit more along with a gif. Let me know what you think. If you would like to see just those changes, it would be commit 44254a42d6429d12975fe39ac3478610a80df619.

**Comment by praneetloke** [comment]
@cnunciato @scottslowe just following-up on this to see if there is anything else I might need to do to move this along? I'll update the post's date in the front-matter based on when this can be merged. Also not sure if you had a chance to the last [comment](https://github.com/pulumi/pulumi-hugo/pull/2397#discussion_r1083088544) from me about the GIF.

**Comment by scottslowe** [comment]
> @cnunciato @scottslowe just following-up on this to see if there is anything else I might need to do to move this along? I'll update the post's date in the front-matter based on when this can be merged. Also not sure if you had a chance to the last [comment](https://github.com/pulumi/pulumi-hugo/pull/2397#discussion_r1083088544) from me about the GIF.

Hi @praneetloke! I was literally reading/reviewing the PR again when I saw an e-mail notification about your comment. I just finished reviewing and I think it's good to go (I want @cnunciato to weigh in as well). For now, we will hold off on merging to publish so that we can put it into our schedule for publishing posts, but no further action (except perhaps adjusting the date) is needed from your end. Thank you for this great post!

**Comment by cnunciato** [comment]
Tentatively scheduled for Feb 6-10.

**Comment by praneetloke** [comment]
@scottslowe wonderful! I've set the date to 2/6 as per @cnunciato's comment. Let me know if you would like me to adjust it again as we get closer to 2/6.

File: themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md
Comment (cnunciato) [inline themes/default/content/blog/organizing-aws-accounts-with-pulumi/index.md:9]:
```suggestion
    - aws
    - enterprise
    - components
```

**Pull request merged**
